compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/internal/ClipboardUtils.desktop.kt

```diff
@@ -63,6 +85,32 @@ internal actual fun ClipEntry?.hasText(): Boolean {
 // Note: the name can't be just `hasText` because NativeClipboard is a typealias to Any,
 // so it would conflict with ClipEntry?.hasText declaration. Therefore, we need a unique name.
 internal fun NativeClipboard.nativeClipboardHasText(): Boolean {
-    val awtClipboard = this as? java.awt.datatransfer.Clipboard ?: return false
+    val awtClipboard = this as? Clipboard ?: return false
     return awtClipboard.isDataFlavorAvailable(DataFlavor.stringFlavor)
-}
\ No newline at end of file
+}
+
+// Derived from StringSelection
+@VisibleForTesting
+internal class AnnotatedStringTransferable(
+    private val data: AnnotatedString
+) : Transferable, ClipboardOwner {
+    override fun getTransferDataFlavors(): Array<DataFlavor?> = supportedFlavors
+
+    override fun isDataFlavorSupported(flavor: DataFlavor): Boolean =
+        flavor in supportedFlavors
+
+    override fun getTransferData(flavor: DataFlavor): Any =
+        when (flavor) {
+            _annotatedStringFlavor -> data
+            DataFlavor.stringFlavor -> data.text
+            else -> throw UnsupportedFlavorException(flavor)
+        }
+
+    override fun lostOwnership(clipboard: Clipboard?, contents: Transferable?) {
+        // Empty
+    }
+
+    companion object {
+        private val supportedFlavors = arrayOf(_annotatedStringFlavor, DataFlavor.stringFlavor)
```

Comment:
Note: when AWT puts this transferable in the clipboard, it drops the `AnnotatedString` flavour, since it's not serializable. The net result is having just the plain text put in the system clipboard, which is consistent with the current behaviour.

Reply:
> it drops the AnnotatedString flavour, since it's not serializable.

What if I have this code **in the same app** (ignoring the fact I use internal functions here):

```kotlin

val annotatedString = AnnotatedString(...)
val clipEntry = annotatedString.toClipEntry()!!
clipboard.setClipEntry(clipEntry)
...

val result = clipboard.getClipEntry().readAnnotatedString()
```

is `result` expected to be equal to `annotatedString` ?

Reply:
Yes, it's what one of the tests is checking, too.

In prod code however the clip entry is handed immediately to the Clipboard, which in turns sets it immediately in the AWT clipboard, which drops the annotated string as described above. So, the only time the annotated string really "exists" in the transferable is while in transit via the Compose clipboard. As soon as it's handed over to AWT it is discarded.